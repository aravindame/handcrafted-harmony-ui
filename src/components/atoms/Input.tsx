import { Form } from 'react-bootstrap'

interface InputProps {
  label: string,
  value: string,
  type: string,
  placeholder: string,
  onChange: (value: string) => void,
  isInvalid?: boolean,
}

const Input = ({ label, value, type, placeholder, onChange, isInvalid }: InputProps) => {
  return (
    <Form.Group className="mb-3" controlId={`formBasic${label}`}>
      <Form.Label>{label}</Form.Label>
      <Form.Control
        type={type}
        placeholder={placeholder}
        value={value}
        onChange={(e) => onChange(e.target.value)}
        isInvalid={isInvalid}
      />
      <Form.Control.Feedback type="invalid">
        Please enter a valid {label.toLowerCase()}.
      </Form.Control.Feedback>
    </Form.Group>
  )
}

export default Input
