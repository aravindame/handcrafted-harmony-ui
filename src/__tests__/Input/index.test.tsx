import { render, screen, fireEvent } from '@testing-library/react'
import Input from '@/components/atoms/Input'

describe('Input component', () => {
  test('renders with correct label and placeholder', () => {
    const label = 'Username'
    const placeholder = 'Enter your username'
    const onChange = jest.fn()

    const component = render(
      <Input
        label={label}
        value=""
        type="text"
        placeholder={placeholder}
        onChange={onChange}
      />
    )

    const inputElement = screen.getByPlaceholderText(placeholder)
    expect(screen.getByLabelText(label)).toBeInTheDocument()
    expect(inputElement).toBeInTheDocument()
    expect(inputElement).toHaveAttribute('type', 'text')
    expect(component).toMatchSnapshot()
  })

  test('calls onChange callback when input value changes', () => {
    const onChange = jest.fn()

    const component = render(
      <Input
        label="Email"
        value=""
        type="email"
        placeholder="Enter your email"
        onChange={onChange}
      />
    )

    const inputElement = screen.getByPlaceholderText('Enter your email')
    fireEvent.change(inputElement, { target: { value: 'test@example.com' } })

    expect(onChange).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalledWith('test@example.com')
    expect(component).toMatchSnapshot()
  })

  test('displays invalid feedback when isInvalid prop is true', () => {
    const label = 'Password'
    const invalidMessage = 'Please enter a valid password.'

    const component = render(
      <Input
        label={label}
        value=""
        type="password"
        placeholder="Enter your password"
        onChange={() => {return}}
        isInvalid
      />
    )

    const invalidFeedback = screen.getByText(invalidMessage)
    expect(invalidFeedback).toBeInTheDocument()
    expect(component).toMatchSnapshot()
  })
})
