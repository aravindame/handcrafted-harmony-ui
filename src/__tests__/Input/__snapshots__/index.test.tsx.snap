// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component calls onChange callback when input value changes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formBasicEmail"
        >
          Email
        </label>
        <input
          class="form-control"
          id="formBasicEmail"
          placeholder="Enter your email"
          type="email"
          value=""
        />
        <div
          class="invalid-feedback"
        >
          Please enter a valid 
          email
          .
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicEmail"
      >
        Email
      </label>
      <input
        class="form-control"
        id="formBasicEmail"
        placeholder="Enter your email"
        type="email"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Please enter a valid 
        email
        .
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input component displays invalid feedback when isInvalid prop is true 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formBasicPassword"
        >
          Password
        </label>
        <input
          class="form-control is-invalid"
          id="formBasicPassword"
          placeholder="Enter your password"
          type="password"
          value=""
        />
        <div
          class="invalid-feedback"
        >
          Please enter a valid 
          password
          .
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicPassword"
      >
        Password
      </label>
      <input
        class="form-control is-invalid"
        id="formBasicPassword"
        placeholder="Enter your password"
        type="password"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Please enter a valid 
        password
        .
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Input component renders with correct label and placeholder 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="formBasicUsername"
        >
          Username
        </label>
        <input
          class="form-control"
          id="formBasicUsername"
          placeholder="Enter your username"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
        >
          Please enter a valid 
          username
          .
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicUsername"
      >
        Username
      </label>
      <input
        class="form-control"
        id="formBasicUsername"
        placeholder="Enter your username"
        type="text"
        value=""
      />
      <div
        class="invalid-feedback"
      >
        Please enter a valid 
        username
        .
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
